File: Console/Programmers/I2C.cs
using FlashMemory;
using System;
using System.Collections.Generic;
using USB;

public class I2C_Programmer : MemoryDeviceUSB
{
    private FCUSB_DEVICE FCUSB;

    public event MemoryDeviceUSB.PrintConsoleEventHandler PrintConsole;
    public event MemoryDeviceUSB.SetProgressEventHandler SetProgress;

    public List<I2C_DEVICE> I2C_EEPROM_LIST;
    private I2C_DEVICE MyFlashDevice = null;

    public I2C_Programmer(FCUSB_DEVICE parent_if) {
        SetProgress?.Invoke(0);
        FCUSB = parent_if;
        I2C_EEPROM_LIST = new List<I2C_DEVICE>();
        I2C_EEPROM_LIST.Add(new I2C_DEVICE("24XX01", 128U, 1, 8));
        I2C_EEPROM_LIST.Add(new I2C_DEVICE("24XX02", 256U, 1, 8));
        I2C_EEPROM_LIST.Add(new I2C_DEVICE("24XX03", 256U, 1, 16));
        I2C_EEPROM_LIST.Add(new I2C_DEVICE("24XX04", 512U, 1, 16));
        I2C_EEPROM_LIST.Add(new I2C_DEVICE("24XX05", 512U, 1, 32));
        I2C_EEPROM_LIST.Add(new I2C_DEVICE("24XX08", 1024U, 1, 16));
        I2C_EEPROM_LIST.Add(new I2C_DEVICE("24XX16", 2048U, 1, 16));
        I2C_EEPROM_LIST.Add(new I2C_DEVICE("24XX32", 4096U, 2, 32));
        I2C_EEPROM_LIST.Add(new I2C_DEVICE("24XX64", 8192U, 2, 32));
        I2C_EEPROM_LIST.Add(new I2C_DEVICE("24XX128", 16384U, 2, 32));
        I2C_EEPROM_LIST.Add(new I2C_DEVICE("24XX256", 32768U, 2, 32));
        I2C_EEPROM_LIST.Add(new I2C_DEVICE("24XX256", 65536U, 2, 32));
        I2C_EEPROM_LIST.Add(new I2C_DEVICE("24XXM01", 131072U, 2, 32));
        I2C_EEPROM_LIST.Add(new I2C_DEVICE("24XXM02", 262144U, 2, 32));
    }

    public void SelectDeviceIndex(int ic2_eeprom_index) {
        MyFlashDevice = I2C_EEPROM_LIST[ic2_eeprom_index];
    }

    public void SelectDeviceIndex(string ic2_eeprom_name) {
        MyFlashDevice = null;
        foreach (var device in I2C_EEPROM_LIST) {
            if (device.NAME.ToUpper().Equals(ic2_eeprom_name.ToUpper())) {
                MyFlashDevice = device;
                return;
            }
        }
    }

    public bool DeviceInit() {
        if (MyFlashDevice is null)
            return false;
        ushort cd_value = (ushort)((int)MainApp.MySettings.I2C_SPEED << 8 | MainApp.MySettings.I2C_ADDRESS); // 02A0
        ushort cd_index = (ushort)((ushort)MyFlashDevice.AddressSize << 8 | (ushort)MyFlashDevice.PAGE_SIZE); // addr size, page size   '0220
        uint config_data = (uint)cd_value << 16 | cd_index;
        bool detect_result = FCUSB.USB_CONTROL_MSG_OUT(USBREQ.I2C_INIT, null, config_data);
        Utilities.Sleep(50); // Wait for IO VCC to charge up
        return detect_result;
    }

    public Device GetDevice {
        get {
            return this.MyFlashDevice;
        }
    }

    public string DeviceName {
        get {
            if (MyFlashDevice is null)
                return "";
            return MyFlashDevice.NAME;
        }
    }

    public long DeviceSize {
        get {
            if (MyFlashDevice is null)
                return 0L;
            return MyFlashDevice.FLASH_SIZE;
        }
    }

    public bool EraseDevice() {
        PrintConsole?.Invoke("I2C ERASE device not supported");
        return false;
    }

    public byte[] ReadData(long flash_offset, long data_count) {
        try {
            var setup_data = new byte[7];
            bool result = false;
            setup_data[0] = (byte)(flash_offset >> 24 & 255L);
            setup_data[1] = (byte)(flash_offset >> 16 & 255L);
            setup_data[2] = (byte)(flash_offset >> 8 & 255L);
            setup_data[3] = (byte)(flash_offset & 255L);
            setup_data[4] = (byte)(data_count >> 16 & 255L);
            setup_data[5] = (byte)(data_count >> 8 & 255L);
            setup_data[6] = (byte)(data_count & 255L);
            var data_out = new byte[(int)(data_count - 1L + 1)];
            result = FCUSB.USB_SETUP_BULKIN(USBREQ.I2C_READEEPROM, setup_data, ref data_out, (uint)data_count);
            if (!result)
                return null;
            if (GetResultStatus() == I2C_STATUS.NOERROR)
                return data_out;
        } catch {
        }
        return null;
    }

    public uint SectorCount()
    {
        return 1U;
    }

    public bool SectorErase(uint SectorIndex)
    {
        return true;
    }

    public long SectorFind(uint SectorIndex)
    {
        return 0L;
    }

    public uint SectorSize(long sector) {
        return (uint)DeviceSize;
    }

    public bool SectorWrite(uint SectorIndex, byte[] data, WriteParameters Params) {
        return false; // Not supported
    }

    public void WaitUntilReady() {
        Utilities.Sleep(100);
    }

    public bool WriteData(long flash_offset, byte[] data_to_write, WriteParameters Params) {
        try {
            var setup_data = new byte[7];
            uint data_count = (uint)data_to_write.Length;
            bool result = false;
            setup_data[0] = (byte)(flash_offset >> 24 & 255L);
            setup_data[1] = (byte)(flash_offset >> 16 & 255L);
            setup_data[2] = (byte)(flash_offset >> 8 & 255L);
            setup_data[3] = (byte)(flash_offset & 255L);
            setup_data[4] = (byte)(data_count >> 16 & 255L);
            setup_data[5] = (byte)(data_count >> 8 & 255L);
            setup_data[6] = (byte)(data_count & 255L);
            result = FCUSB.USB_SETUP_BULKOUT(USBREQ.I2C_WRITEEEPROM, setup_data, data_to_write, data_count);
            if (!result)
                return false;
            FCUSB.USB_WaitForComplete(); // It may take a few microseconds to complete
            if (GetResultStatus() == I2C_STATUS.NOERROR)
                return true;
        } catch {
        }
        return false;
    }

    public uint SectorSize(uint sector) {
        throw new NotImplementedException();
    }

    private I2C_STATUS GetResultStatus() {
        try {
            var packet_out = new byte[1];
            if (!FCUSB.USB_CONTROL_MSG_IN(USBREQ.I2C_RESULT, ref packet_out))
                return I2C_STATUS.USBFAIL;
            return (I2C_STATUS)packet_out[0];
        } catch {
            return I2C_STATUS.ERROR;
        }
    }

    private enum I2C_STATUS : byte {
        USBFAIL = 0,
        NOERROR = 0x50,
        ERROR = 0x51
    }

    internal bool WriteData(long address, byte[] packet_data) {
        throw new NotImplementedException();
    }

    internal bool IsConnected() {
        throw new NotImplementedException();
    }
}

File: FlashcatUSB/I2C.vb
Imports FlashcatUSB.FlashMemory
Imports FlashcatUSB.USB.HostClient

Public Class I2C_Programmer : Implements MemoryDeviceUSB
    Private FCUSB As FCUSB_DEVICE

    Public Event PrintConsole(message As String) Implements MemoryDeviceUSB.PrintConsole
    Public Event SetProgress(percent As Integer) Implements MemoryDeviceUSB.SetProgress

    Private eeprom_size As UInt32 = 0

    Sub New(ByVal parent_if As FCUSB_DEVICE)
        FCUSB = parent_if
        Create_I2C_EEPROM_List()
    End Sub

    Public Function DeviceInit() As Boolean Implements MemoryDeviceUSB.DeviceInit
        eeprom_size = MySettings.I2C_SIZE
        Dim page_size As Byte
        Dim addr_size As Byte
        GetAddressPageSize(addr_size, page_size)
        Dim cd_value As UInt16 = (CUShort(MySettings.I2C_SPEED) << 8) Or (MySettings.I2C_ADDRESS) '02A0
        Dim cd_index As UInt16 = (CUShort(addr_size) << 8) Or (page_size) 'addr size, page size   '0220
        Dim config_data As UInt32 = (CUInt(cd_value) << 16) Or cd_index
        Dim detect_result As Boolean = FCUSB.USB_CONTROL_MSG_OUT(USB.USBREQ.I2C_INIT, Nothing, config_data)
        Return detect_result
    End Function

    Public Class I2C_DEVICE
        Public ReadOnly Property Name As String
        Public ReadOnly Property Size As UInt32 'Number of bytes in this Flash device
        Public ReadOnly Property AddressSize As Integer
        Public ReadOnly Property PageSize As Integer

        Sub New(ByVal DisplayName As String, ByVal SizeInBytes As UInt32, ByVal EEAddrSize As Integer, ByVal EEPageSize As Integer)
            Me.Name = DisplayName
            Me.Size = SizeInBytes
            Me.AddressSize = EEAddrSize 'Number of bytes that are used to store the address
            Me.PageSize = EEPageSize
        End Sub

    End Class

    Public I2C_EEPROM_LIST As New List(Of I2C_DEVICE)

    Public ReadOnly Property DeviceName As String Implements MemoryDeviceUSB.DeviceName
        Get
            Return GetConfiguredDeviceName(Me.DeviceSize)
        End Get
    End Property

    Public ReadOnly Property DeviceSize As Long Implements MemoryDeviceUSB.DeviceSize
        Get
            Return eeprom_size
        End Get
    End Property

    Public ReadOnly Property SectorSize(sector As UInteger, Optional area As FlashArea = FlashArea.Main) As UInteger Implements MemoryDeviceUSB.SectorSize
        Get
            Return Me.DeviceSize
        End Get
    End Property

    Private Sub Create_I2C_EEPROM_List()
        I2C_EEPROM_LIST.Add(New I2C_DEVICE("24XX01", 128, 1, 8))
        I2C_EEPROM_LIST.Add(New I2C_DEVICE("24XX02", 256, 1, 8))
        I2C_EEPROM_LIST.Add(New I2C_DEVICE("24XX04", 512, 1, 16))
        I2C_EEPROM_LIST.Add(New I2C_DEVICE("24XX08", 1024, 1, 16))
        I2C_EEPROM_LIST.Add(New I2C_DEVICE("24XX16", 2048, 1, 16))
        I2C_EEPROM_LIST.Add(New I2C_DEVICE("24XX32", 4096, 2, 32))
        I2C_EEPROM_LIST.Add(New I2C_DEVICE("24XX64", 8192, 2, 32))
        I2C_EEPROM_LIST.Add(New I2C_DEVICE("24XX128", 16384, 2, 32))
        I2C_EEPROM_LIST.Add(New I2C_DEVICE("24XX256", 32768, 2, 32))
        I2C_EEPROM_LIST.Add(New I2C_DEVICE("24XX256", 65536, 2, 32))
        I2C_EEPROM_LIST.Add(New I2C_DEVICE("24XXM01", 131072, 2, 32))
        I2C_EEPROM_LIST.Add(New I2C_DEVICE("24XXM02", 262144, 2, 32))
    End Sub

    Public Function IsConnected() As Boolean
        Try
            Me.DeviceInit()
            Utilities.Sleep(20) 'Wait for IO VCC to charge up
            Dim test_data() As Byte = Me.ReadData(0, 16) 'This does a test read to see if data is read
            If test_data Is Nothing Then Return False
            Return True
        Catch ex As Exception
        End Try
        Return False
    End Function

    Public Function GetConfiguredDeviceName(ByVal device_size As Integer) As String
        Select Case device_size
            Case 128
                Return "24XX01"
            Case 256
                Return "24XX02"
            Case 512
                Return "24XX04"
            Case 1024
                Return "24XX08"
            Case 2048
                Return "24XX16"
            Case 4096
                Return "24XX32"
            Case 8192
                Return "24XX64"
            Case 16384
                Return "24XX128"
            Case 32768
                Return "24XX256"
            Case 65536
                Return "24XX512"
            Case 131072
                Return "M24M01"
            Case 262144
                Return "M24M02"
            Case Else
                Return ""
        End Select
    End Function

    Private Sub GetAddressPageSize(ByRef addr_size As Byte, ByRef page_size As Byte)
        Select Case MySettings.I2C_SIZE
            Case 128
                addr_size = 1
                page_size = 8
            Case 256
                addr_size = 1
                page_size = 8
            Case 512
                addr_size = 1
                page_size = 16
            Case 1024
                addr_size = 1
                page_size = 16
            Case 2048
                addr_size = 1
                page_size = 16
            Case 4096
                addr_size = 2
                page_size = 32
            Case 8192
                addr_size = 2
                page_size = 32
            Case 16384
                addr_size = 2
                page_size = 32
            Case 32768
                addr_size = 2
                page_size = 32
            Case 65536
                addr_size = 2
                page_size = 32
            Case 131072
                addr_size = 2
                page_size = 32
            Case 262144
                addr_size = 2
                page_size = 32
        End Select
    End Sub

    Private Function GetResultStatus() As I2C_STATUS
        Try
            Dim packet_out(0) As Byte
            If Not FCUSB.USB_CONTROL_MSG_IN(USB.USBREQ.I2C_RESULT, packet_out) Then Return I2C_STATUS.USBFAIL
            Return packet_out(0)
        Catch ex As Exception
            Return I2C_STATUS.ERROR
        End Try
    End Function

    Private Enum I2C_STATUS As Byte
        USBFAIL = 0
        NOERROR = &H50
        [ERROR] = &H51
    End Enum

    Public Function ReadData(flash_offset As Long, data_count As UInteger, Optional area As FlashArea = FlashArea.Main) As Byte() Implements MemoryDeviceUSB.ReadData
        Try
            Dim setup_data(6) As Byte
            Dim result As Boolean = False
            setup_data(0) = ((flash_offset >> 24) And 255)
            setup_data(1) = ((flash_offset >> 16) And 255)
            setup_data(2) = ((flash_offset >> 8) And 255)
            setup_data(3) = (flash_offset And 255)
            setup_data(4) = ((data_count >> 16) And 255)
            setup_data(5) = ((data_count >> 8) And 255)
            setup_data(6) = (data_count And 255)
            Dim data_out(data_count - 1) As Byte
            result = FCUSB.USB_SETUP_BULKIN(USB.USBREQ.I2C_READEEPROM, setup_data, data_out, data_count)
            If Not result Then Return Nothing
            If GetResultStatus() = I2C_STATUS.NOERROR Then Return data_out
        Catch ex As Exception
        End Try
        Return Nothing
    End Function

    Public Function WriteData(flash_offset As Long, data_to_write() As Byte, Optional ByRef Params As WriteParameters = Nothing) As Boolean Implements MemoryDeviceUSB.WriteData
        Try
            Dim setup_data(6) As Byte
            Dim data_count As UInt32 = data_to_write.Length
            Dim result As Boolean = False
            setup_data(0) = ((flash_offset >> 24) And 255)
            setup_data(1) = ((flash_offset >> 16) And 255)
            setup_data(2) = ((flash_offset >> 8) And 255)
            setup_data(3) = (flash_offset And 255)
            setup_data(4) = ((data_count >> 16) And 255)
            setup_data(5) = ((data_count >> 8) And 255)
            setup_data(6) = (data_count And 255)
            result = FCUSB.USB_SETUP_BULKOUT(USB.USBREQ.I2C_WRITEEEPROM, setup_data, data_to_write, data_count)
            If Not result Then Return False
            FCUSB.USB_WaitForComplete() 'It may take a few microseconds to complete
            If GetResultStatus() = I2C_STATUS.NOERROR Then Return True
        Catch ex As Exception
        End Try
        Return False
    End Function

    Public Function EraseDevice() As Boolean Implements MemoryDeviceUSB.EraseDevice
        Return True 'EEPROM does not support erase commands
    End Function

    Public Sub WaitUntilReady() Implements MemoryDeviceUSB.WaitUntilReady
        Utilities.Sleep(10)
    End Sub

    Public Function SectorFind(SectorIndex As UInteger, Optional area As FlashArea = FlashArea.Main) As Long Implements MemoryDeviceUSB.SectorFind
        Return 0
    End Function

    Public Function Sector_Erase(SectorIndex As UInteger, Optional area As FlashArea = FlashArea.Main) As Boolean Implements MemoryDeviceUSB.Sector_Erase
        Return True
    End Function

    Public Function Sector_Count() As UInteger Implements MemoryDeviceUSB.Sector_Count
        Return 1
    End Function

    Public Function Sector_Write(SectorIndex As UInteger, data() As Byte, Optional ByRef Params As WriteParameters = Nothing) As Boolean Implements MemoryDeviceUSB.Sector_Write
        Return False 'Not supported
    End Function

End Class
